src/sandstorm/util.c++

```diff
@@ -95,6 +96,47 @@ kj::Maybe<kj::AutoCloseFd> raiiOpenAtIfExists(
   }
 }
 
+// scratch buffer for file paths; used by raiiOpenAtIfExistsContained:
+thread_local char path_buf[PATH_MAX+1];
+
+kj::Maybe<kj::AutoCloseFd> raiiOpenAtIfExistsContained(int dirfd, kj::StringPtr name, int flags, mode_t mode) {
+  auto path = kj::Path::parse(name);
+  int parentFd = dirfd;
+  int fd;
+  KJ_SYSCALL(fd = dup(dirfd));
+  kj::AutoCloseFd file(fd);
+
+  int symlink_limit = 16; // arbitrary limit
+
+  for(int i = 0; i < path.size(); i++) {
+    const char *part = path[i].cStr();
+    KJ_SYSCALL_HANDLE_ERRORS(fd = openat(file.get(), part, flags | O_NOFOLLOW, mode)) {
+      case ENOENT:
+        return nullptr;
+      case ELOOP:
+        {
+          if(symlink_limit == 0) {
+            KJ_FAIL_SYSCALL("openat()", error);
+          }
+          symlink_limit--;
+
+          memset(&path_buf, 0, sizeof path_buf);
```

Comment:
IMO it'd be cleaner to use `readlinkat()`'s return value to determine where to put the NUL terminator, rather than zero-initialize the buffer upfront and rely on that. That way does less work and also gives you a chance to check if the link data turned out to be longer than PATH_MAX (which is very possible, there's no actual limit on path lengths, PATH_MAX exists for historical reasons).

You could also use the KJ filesystem API's readlink() which automatically grows the buffer as needed to fit the full path.

Reply:
Ah, I'd remembered `PATH_MAX` as being the maximum size the kernel will accept as an argument to a syscall (obviously with the way mount() works there's no way to have a limit on full paths), but looking it up again it looks like that's only for syscalls that take a nul-terminated string and no buffer length, so it doesn't apply here. good catch.

Reply:
Done.